pkg/apis/backupdriver/v1/snapshot.go

```diff
@@ -0,0 +1,148 @@
+package v1
+
+import meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SnapshotSpec struct {
+	// ResourceHandle refers to a Kubernetes resource, currently a PVC but this may be extended in the future
+	ResourceHandle string
+
+	// The backup repository to snapshot into.  The namespace the Snapshot/PVC lives in must have access to the repository
+	BackupRepository string
+}
+
+// SnapshotPhase represents the lifecycle phase of a Snapshot.
+// New - No work yet, next phase is InProgress
+// InProgress - snapshot being taken
+// Snapshotted - local snapshot complete, next phase is Protecting or SnapshotFailed
+// SnapshotFailed - end state, snapshot was not able to be taken
+// Uploading - snapshot is being moved to durable storage
+// Uploaded - end state, snapshot has been protected
+// UploadFailed - end state, unable to move to durable storage
+// Canceling - when the SanpshotCancel flag is set, if the Snapshot has not already moved into a terminal state, the
+//             status will move to Canceling.  The snapshot ID will be removed from the status status if has been filled in
+//             and the snapshot ID will not longer be valid for a Clone operation
+// Canceled - the operation was canceled, the snapshot ID is not valid
+type SnapshotPhase string
+
+const (
+	SnapshotPhaseNew            SnapshotPhase = "New"
+	SnapshotPhaseInProgress     SnapshotPhase = "InProgress"
+	SnapshotPhaseSnapshotted    SnapshotPhase = "Snapshotted"
+	SnapshotPhaseSnapshotFailed SnapshotPhase = "SnapshotFailed"
+	SnapshotPhaseUploading      SnapshotPhase = "Uploading"
+	SnapshotPhaseUploaded       SnapshotPhase = "Uploaded"
+	SnapshotPhaseUploadFailed   SnapshotPhase = "UploadFailed"
+	SnapshotPhaseCanceling      SnapshotPhase = "Canceling"
+	SnapshotPhaseCanceled       SnapshotPhase = "Canceled"
+)
```

Comment:
Does "Uploaded" also mean the local snapshot is deleted?

Reply:
Maybe, maybe not.  Currently yes, but not required.  We could make it "durable"